python/lsst/obs/base/formatters/filter.py

```diff
@@ -0,0 +1,151 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("FilterFormatter",)
+
+import yaml
+
+from typing import (
+    Any,
+    Optional,
+    Type,
+)
+
+from lsst.daf.butler.formatters.file import FileFormatter
+
+
+class FilterFormatter(FileFormatter):
+    """Read and write `~lsst.afw.image.Filter` filter information."""
+
+    extension = ".yaml"
+
+    unsupportedParameters = None
+    """This formatter does not support any parameters."""
+
+    def _readFile(self, path: str, pytype: Type[Any] = None) -> Any:
+        """Read a file from the path in YAML format.
+
+        Parameters
+        ----------
+        path : `str`
+            Path to use to open YAML format file.
+        pytype : `class`, optional
+            Not used by this implementation.
+
+        Returns
+        -------
+        data : `object`
+            Either data as Python object read from YAML file, or None
+            if the file could not be opened.
+
+        Notes
+        -----
+        The `~yaml.UnsafeLoader` is used when parsing the YAML file.
+        """
+        try:
+            with open(path, "rb") as fd:
+                data = self._fromBytes(fd.read(), pytype)
+        except FileNotFoundError:
+            data = None
+
+        return data
+
+    def _fromBytes(self, serializedDataset: bytes, pytype: Optional[Type[Any]] = None) -> Any:
+        """Read the bytes object as a python object.
+
+        Parameters
+        ----------
+        serializedDataset : `bytes`
+            Bytes object to unserialize.
+        pytype : `class`, optional
+            Expected python type to be returned.
+
+        Returns
+        -------
+        inMemoryDataset : `object`
+            The requested data as an object, or None if the string could
+            not be read.
+        """
+        data = yaml.load(serializedDataset, Loader=yaml.FullLoader)
+
+        # This will be a simple dict so we need to convert it to
+        # the Filter type -- just needs the name
+        filter = pytype(data["canonicalName"])
+
+        return filter
+
+    def _writeFile(self, inMemoryDataset: Any) -> None:
+        """Write the in memory dataset to file on disk.
+
+        Will look for `_asdict()` method to aid YAML serialization, following
+        the approach of the simplejson module.
+
+        Parameters
+        ----------
+        inMemoryDataset : `object`
+            Object to serialize.
+
+        Raises
+        ------
+        Exception
+            The file could not be written.
+        """
+        with open(self.fileDescriptor.location.path, "wb") as fd:
+            fd.write(self._toBytes(inMemoryDataset))
+
+    def _toBytes(self, inMemoryDataset: Any) -> bytes:
+        """Write the in memory dataset to a bytestring.
+
+        Parameters
+        ----------
+        inMemoryDataset : `object`
```

Comment:
This is a bit contradictory, and I'm not sure what to do about it. On the one hand, formal type safety requires that this be an object of any type. On the other hand, the method body explicitly assumes that `inMemoryDataset` is a `Filter`, and will raise an `AttributeError` (convert to `TypeError`?) for any other type.

Reply:
We haven't yet developed a convention for someone configuring their datastore such that the wrong python type turns up in the formatter.